File: packages/python/elizaos/bootstrap/actions/unfollow_room.py
from __future__ import annotations

from dataclasses import dataclass, field
from typing import TYPE_CHECKING

from elizaos.types import Action, ActionExample, ActionResult, Content, MemoryType

if TYPE_CHECKING:
    from elizaos.types import HandlerCallback, HandlerOptions, IAgentRuntime, Memory, State


@dataclass
class UnfollowRoomAction:
    name: str = "UNFOLLOW_ROOM"
    similes: list[str] = field(
        default_factory=lambda: [
            "LEAVE_ROOM",
            "UNSUBSCRIBE_ROOM",
            "STOP_WATCHING_ROOM",
            "EXIT_ROOM",
        ]
    )
    description: str = (
        "Stop following a room and cease receiving updates. "
        "Use this when you no longer want to monitor a room's activity."
    )

    async def validate(
        self, runtime: IAgentRuntime, message: Memory, _state: State | None = None
    ) -> bool:
        room_id = message.room_id
        if not room_id:
            return False

        room = await runtime.get_room(room_id)
        if room is None:
            return False

        world_id = room.world_id
        if world_id:
            world = await runtime.get_world(world_id)
            if world and world.metadata:
                followed_rooms = world.metadata.get("followedRooms", [])
                if str(room_id) in followed_rooms:
                    return True

        return False

    async def handler(
        self,
        runtime: IAgentRuntime,
        message: Memory,
        state: State | None = None,
        options: HandlerOptions | None = None,
        callback: HandlerCallback | None = None,
        responses: list[Memory] | None = None,
    ) -> ActionResult:
        room_id = message.room_id
        if not room_id:
            return ActionResult(
                text="No room specified to unfollow",
                values={"success": False, "error": "no_room_id"},
                data={"actionName": "UNFOLLOW_ROOM"},
                success=False,
            )

        room = await runtime.get_room(room_id)
        if room is None:
            return ActionResult(
                text="Room not found",
                values={"success": False, "error": "room_not_found"},
                data={"actionName": "UNFOLLOW_ROOM"},
                success=False,
            )

        room_name = str(room.name) if room.name else "Unknown Room"

        world_id = room.world_id
        if world_id:
            world = await runtime.get_world(world_id)
            if world and world.metadata:
                followed_rooms = list(world.metadata.get("followedRooms", []))
                room_id_str = str(room_id)

                if room_id_str in followed_rooms:
                    followed_rooms.remove(room_id_str)
                    world.metadata["followedRooms"] = followed_rooms
                    await runtime.update_world(world)

        await runtime.create_memory(
            content=Content(
                text=f"Stopped following room: {room_name}",
                actions=["UNFOLLOW_ROOM"],
            ),
            room_id=room_id,
            entity_id=runtime.agent_id,
            memory_type=MemoryType.ACTION,
            metadata={"type": "UNFOLLOW_ROOM", "roomName": room_name},
        )

        response_content = Content(
            text=f"I am no longer following {room_name}.",
            actions=["UNFOLLOW_ROOM"],
        )

        if callback:
            await callback(response_content)

        return ActionResult(
            text=f"Stopped following room: {room_name}",
            values={
                "success": True,
                "unfollowed": True,
                "roomId": str(room_id),
                "roomName": room_name,
            },
            data={
                "actionName": "UNFOLLOW_ROOM",
                "roomId": str(room_id),
                "roomName": room_name,
            },
            success=True,
        )

    @property
    def examples(self) -> list[list[ActionExample]]:
        return [
            [
                ActionExample(
                    name="{{name1}}",
                    content=Content(text="You can stop watching this channel now."),
                ),
                ActionExample(
                    name="{{name2}}",
                    content=Content(
                        text="I'll stop following this room.",
                        actions=["UNFOLLOW_ROOM"],
                    ),
                ),
            ],
        ]


unfollow_room_action = Action(
    name=UnfollowRoomAction.name,
    similes=UnfollowRoomAction().similes,
    description=UnfollowRoomAction.description,
    validate=UnfollowRoomAction().validate,
    handler=UnfollowRoomAction().handler,
    examples=UnfollowRoomAction().examples,
)

File: packages/rust/src/bootstrap/actions/unfollow_room.rs
//! UNFOLLOW_ROOM action implementation.

use async_trait::async_trait;
use std::collections::HashMap;
use std::sync::Arc;

use crate::error::{PluginError, PluginResult};
use crate::runtime::IAgentRuntime;
use crate::types::{ActionResult, Content, Memory, MemoryType, State};

use super::Action;

/// Action for unfollowing a room.
pub struct UnfollowRoomAction;

#[async_trait]
impl Action for UnfollowRoomAction {
    fn name(&self) -> &'static str {
        "UNFOLLOW_ROOM"
    }

    fn similes(&self) -> &[&'static str] {
        &[
            "LEAVE_ROOM",
            "UNSUBSCRIBE_ROOM",
            "STOP_WATCHING_ROOM",
            "EXIT_ROOM",
        ]
    }

    fn description(&self) -> &'static str {
        "Stop following a room and cease receiving updates. \
         Use this when you no longer want to monitor a room's activity."
    }

    async fn validate(&self, runtime: &dyn IAgentRuntime, message: &Memory) -> bool {
        let Some(room_id) = message.room_id else {
            return false;
        };

        // Check if room exists
        let Ok(Some(room)) = runtime.get_room(room_id).await else {
            return false;
        };

        // Check if currently following
        if let Some(world_id) = room.world_id {
            if let Ok(Some(world)) = runtime.get_world(world_id).await {
                if let Some(followed) = world.metadata.get("followedRooms") {
                    if let Some(arr) = followed.as_array() {
                        let room_str = room_id.to_string();
                        if arr.iter().any(|v| v.as_str() == Some(&room_str)) {
                            return true; // Currently following
                        }
                    }
                }
            }
        }

        false
    }

    async fn handler(
        &self,
        runtime: Arc<dyn IAgentRuntime>,
        message: &Memory,
        _state: Option<&State>,
        _responses: Option<&[Memory]>,
    ) -> PluginResult<ActionResult> {
        let room_id = message.room_id.ok_or_else(|| {
            PluginError::InvalidInput("No room specified to unfollow".to_string())
        })?;

        let room = runtime
            .get_room(room_id)
            .await?
            .ok_or_else(|| PluginError::NotFound("Room not found".to_string()))?;

        let room_name = room
            .name
            .clone()
            .unwrap_or_else(|| "Unknown Room".to_string());

        // Update world's followed rooms
        if let Some(world_id) = room.world_id {
            if let Some(mut world) = runtime.get_world(world_id).await? {
                let mut followed: Vec<String> = world
                    .metadata
                    .get("followedRooms")
                    .and_then(|v| v.as_array())
                    .map(|arr| {
                        arr.iter()
                            .filter_map(|v| v.as_str().map(String::from))
                            .collect()
                    })
                    .unwrap_or_default();

                let room_str = room_id.to_string();
                followed.retain(|r| r != &room_str);
                world
                    .metadata
                    .insert("followedRooms".to_string(), serde_json::json!(followed));
                runtime.update_world(&world).await?;
            }
        }

        // Create memory of the action
        let mut metadata = HashMap::new();
        metadata.insert("type".to_string(), serde_json::json!("UNFOLLOW_ROOM"));
        metadata.insert("roomName".to_string(), serde_json::json!(&room_name));

        runtime
            .create_memory(
                Content {
                    text: format!("Stopped following room: {}", room_name),
                    actions: vec!["UNFOLLOW_ROOM".to_string()],
                    ..Default::default()
                },
                Some(room_id),
                Some(runtime.agent_id()),
                MemoryType::Action,
                metadata,
            )
            .await?;

        Ok(
            ActionResult::success(format!("Stopped following room: {}", room_name))
                .with_value("success", true)
                .with_value("unfollowed", true)
                .with_value("roomId", room_id.to_string())
                .with_value("roomName", room_name.clone())
                .with_data("actionName", "UNFOLLOW_ROOM")
                .with_data("roomId", room_id.to_string())
                .with_data("roomName", room_name),
        )
    }
}
